// lucia-sveltekit/client.js
import { getContext } from "svelte";
import { LuciaError } from "./utils/error.js";
export { default as Lucia } from "./Lucia.svelte";
export const signOut = async (accessToken) => {
    const response = await fetch("/api/auth/logout", {
        method: "POST",
        headers: {
            Authorization: `Bearer ${accessToken || ""}`,
        },
    });
    if (response.ok)
        return { error: null };
    let result;
    try {
        result = await response.json();
    }
    catch (e) {
        console.error(e);
        throw new LuciaError("UNKNOWN_ERROR");
    }
    if (result.message)
        throw new LuciaError(result.message);
};
export const refreshTokens = async (refreshToken) => {
    const response = await fetch("/api/auth/refresh", {
        method: "POST",
        headers: {
            Authorization: `Bearer ${refreshToken}`,
        },
    });
    if (!response.ok) {
        let result;
        try {
            result = await response.json();
        }
        catch (e) {
            console.error(e);
            throw new LuciaError("UNKNOWN_ERROR");
        }
        throw new LuciaError(result.message);
    }
    const result = await response.json();
    return {
        refresh_token: result.refresh_token,
        access_token: result.access_token,
    };
};
export const getSession = () => {
    const stores = getContext("__lucia__");
    return stores.session;
};
